sklearn/datasets/openml.py

```diff
@@ -0,0 +1,495 @@
+
+    cache : boolean, default=True
+        Whether to cache downloaded datasets using joblib.
+
+    Returns
+    -------
+
+    data : Bunch
+        Dictionary-like object, the interesting attributes are:
+        'data' (np.array), the data to learn, 'target' (np.array), the
+        regression target or classification labels, 'DESCR' (str), the full
+        description of the dataset, 'feature_names' (list), the original names
+        of the dataset columns, and 'details' (json) which provide more
+        information on the openml meta-data. Missing values in the 'data' and
+        'target' field are represented as NaN's.
+    """
+    data_home = get_data_home(data_home=data_home)
+    data_home = join(data_home, 'openml')
+    if cache:
+        mem = Memory(join(data_home, 'cache'), verbose=0).cache
+    else:
+        def mem(func):
+            return func
+    cached_get_data_info_by_name = mem(_get_data_info_by_name)
+    cached_get_data_description_by_id = mem(_get_data_description_by_id)
+    cached_get_data_features = mem(_get_data_features)
+    cached_download_data_arff = mem(_download_data_arff)
+
+    if not exists(data_home):
+        os.makedirs(data_home)
+
+    # check valid function arguments. data_id XOR (name, version) should be
+    # provided
+    if name is not None:
+        # OpenML is case-insensitive, but the caching mechanism is not
+        # convert all data names (str) to lower case
+        name = name.lower()
+        if data_id is not None:
+            raise ValueError(
+                "Dataset data_id={} and name={} passed, but you can only "
+                "specify a numeric data_id or a name, not "
+                "both.".format(data_id, name))
+        data_info = cached_get_data_info_by_name(name, version)
+        data_id = data_info['did']
+    elif data_id is not None:
+        # from the previous if statement, it is given that name is None
+        if version is not "active":
+            raise ValueError(
+                "Dataset data_id={} and version={} passed, but you can only "
+                "specify a numeric data_id or a version, not "
+                "both.".format(data_id, name))
+    else:
+        raise ValueError(
+            "Neither name nor data_id are provided. Please provide name or "
+            "data_id.")
+
+    data_description = cached_get_data_description_by_id(data_id)
+    if data_description['status'] != "active":
+        warn("Version {} of dataset {} is inactive, meaning that issues have "
+             "been found in the dataset. Try using a newer version from "
+             "this URL: {}".format(
+                data_description['version'],
+                data_description['name'],
+                data_description['url']))
+
+    # download data features, meta-info about column types
+    features_list = cached_get_data_features(data_id)
+    features_dict = {feature['name']: feature for feature in features_list}
```

Comment:
The switching back and forth between using `features_dict` and `features_list` is confusing. It would be better if `features_dict` were only defined close to where it's used.

Reply:
Seems subjective to me, but I'm fine with moving it down